### ISSUE TITLE
[K-Improvement] K Developer Container for Easier Installation

### ISSUE THREAD

**Issue by sskeirik** [runtimeverification/k#2886]
### Motivation

**TLDR:** To make K development accessible beyond RV itself, we need a packaging solution that just works, in the sense that all optional features are either: (a) already enabled; or (b) can consistently and easily enabled.
I propose that we produce a K dev container where all K features just work that operates on host-mounted directories.

**Full version:** To make K development accessible beyond RV itself, we need a packaging solution where all features of K just works.
This is not just a theoretical issue: we have clients who are interested in adopting K but who find installing K across their diverse dev environments to be a roadblock, esp. those with many developers on Arm macOS.

Currently, reliably bringing up a K environment in which all features of K just work across multiple OSes is hard, especially for rolling release distributions (e.g. macOS/brew, Arch Linux) since K is sensitive to certain package versions, e.g., LLVM and z3.
Additionally, certain K features rely on packages that are not dependencies of K itself, e.g., debugging support and the crypto hooks.

In order to have a one-size-fits-all K toolchain, I propose we extend our existing K docker container to provide a K dev container solution such that:

1.  in addition to including K, the K dev container image would include correct versions of all ancilliary packages (e.g. z3, crypto libs, LLVM, etc.) needed to enable the entire K stlib plus all K features
2.  we write new versions of the K entrypoint scripts, i.e., `kompile`, `krun`, `kprove`, etc..., to be shipped with the container, which work by mounting a host directory into the container and running the container-installed K tooling
3.  until Arm macOS is supported fully, we also provide a wrapper script, e.g., to bring up a QEMU x86_64 VM when needed or similar, so that all K tooling just works

I believe this approach could:

1.  provide a more convenient and consistent K tooling experience across different distributions;
2.  possibly provide a simpler K application installation method;
3.  possibly provide a simpler, more unified, K development/CI experience and reduce CI load.

To elaborate briefly on the second point:

-   shipping K-based applications can be a pain
-   we could install K applications as docker containers layered on top of the standard K dev container
-   this would ensure that the K application could be run _and_ further developed in the container

To elaborate briefly on the third point:

-   most K applications use similar cryptography libraries and or other tools at specific versions (e.g. z3)
-   we could just install all of the common tools and crypto libraries in our K dev container, which will be cached, instead of rebuilding/reinstalling them on every CI run
-   using the same K dev container for shipping applications and for CI increases reproducibility and minimizes CI retries

### Example K Code

N/A

### Documentation

This feature would require more design work before proper documentation could be written.

### Potential Alternatives/Workarounds

The alternative is what we already do: ship K tools and K-based applications, typically, as OS distribution packages.

The issue with this is that there are so many different packages, each with different different packaged versions, so it's very hard to write a one-size-fits-all packaged solution.

### Testing Approach

This feature would require more design work before proper tests could be written.

**Comment by goodlyrottenapple** [comment]
many of these are already addressed by our nix packaging, e.g. whilst still flaky, nix is currently (probably) the least painful way to get K running on M1 macs. Also you can run

```bash
nix shell github:runtimeverification/k
```

and immediately useK  tools such as `kast`/`kompile` in your current shell, without having to globally install K. We also keep track of exactly which version of Z3 we use in nix and it doesn't interfere with any version the user might have globally installed.

**Comment by ehildenb** [comment]
We don't want to work on this until the Nix stuff is ironed out.

**Comment by Baltoli** [comment]
We have `kup` now which - broadly speaking - addresses the issue of external users building tools that use K. If there's a future use case that `kup` doesn't handle, we can open a new issue.
